src/Ast.ml

```diff
@@ -647,6 +647,7 @@ end = struct
       | Pexp_setfield (_, _, e0) when e0 == exp -> Some (LessMinus, Non)
       | Pexp_setinstvar _ -> Some (LessMinus, Non)
       | Pexp_field _ -> Some (Dot, Left)
+      | Pexp_send _ -> Some (HashOp, Left)
```

Comment:
Isn't `#` non-associative?

Reply:
My understanding of this <http://caml.inria.fr/pub/docs/manual-ocaml/expr.html> is that it *behaves* like `.`.

Also it was already the decision there: https://github.com/smondet/ocamlformat/blob/pexp-send/src/Ast.ml#L171-L172 :)

Testing does not throw ast-diff-errors for all these:

```ocaml
let f x = (x ()).contents
let f x = (x ()).contents#method_of_ref ()
let f x = ((x ()).contents ())#method_of_ref ()
let f x = x#hello.contents.contents
let f x = x#hello.contents#bye.contents
let f x = x.contents#hello
let f x = x.contents#hello.contents
```

Reply:
I'm not sure whether the manual is referring to the `#` send expression or the prefix ops that start with a `#` char, but the [parser](https://github.com/ocaml/ocaml/blob/trunk/parsing/parser.mly#L598) seems to treat them differently.

But you may be right that once code parses, then it might not be possible to trigger a formatting error. I don't know, but would rather match the parser if possible.

Reply:
With `Non` instead of `Left` for `HashOp` it gives something like this:

```oaml
let f x = (x ()).contents

let f x = (x ()).contents#method_of_ref ()

let f x = ((x ()).contents ())#method_of_ref ()

let f x = (x#hello).contents.contents

let f x = ((x#hello).contents#bye).contents

let f x = x.contents#hello

let f x = (x.contents#hello).contents
````

My eyes/brain are trained to find parentheses helpful in those cases but I know it's not everyone's taste :)

Tell me if you prefer with `Non` and I'll commit that.

(PS the ops that start with `#` are not supposed to be part of the language itself, they are REPL-only, so in another chapter:
<http://caml.inria.fr/pub/docs/manual-ocaml/toplevel.html> 
There is still another `#` for type definitions and for pattern mattching though ... OCaml syntax :heart: )

Reply:
I think that the extra parens are due to the current order of `Dot` and `HashOp` being wrong in the definition of `type prec`. Fixing that and going with `Non`-associative seems like the right option.

(I think I saw somewhere that the  `#` ops were intended/allowed for syntax extensions, not just the repl directives. Not sure. But yeah, the OCaml syntax is not exactly small...)

Reply:
(Ah right there is the `js_of_ocaml` ppx thing that converts `##` and `##.`!)

OK I'm changing the order in `pred` also :+1: